backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -0,0 +1,258 @@
+        /// <exception cref="KeyNotFoundException">Project report does not exist in the datasource.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to view snapshots.</exception>
+        /// <returns></returns>
+        public IEnumerable<ProjectSnapshot> GetSnapshots(int reportId)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var report = this.Context.ProjectReports
+                .FirstOrDefault(p => p.Id == reportId) ?? throw new KeyNotFoundException();
+
+            return GetSnapshots(report);
+        }
+
+        /// <summary>
+        /// Add the specified report to the datasource.
+        /// The 'From' date is set to the 'To' date of the most recent report, if one exists.
+        /// For every SPL project in the system, a new Snapshot is generated using current project data and a 'snapshotOn' time equal to the newly created reports 'To' date.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to add report.</exception>
+        /// <returns></returns>
+        public ProjectReport Add(ProjectReport report)
+        {
+            if (report.To == null)
+            {
+                report.To = DateTime.UtcNow;
+            }
+            var mostRecentReport = this.Context.ProjectReports
+                .Where(r => r.ReportTypeId == ReportTypes.SPL)
+                .OrderByDescending(r => r.To)
+                .FirstOrDefault();
+            report.From = mostRecentReport?.To;
+            this.Context.Add(report);
+
+            this.Context.ProjectSnapshots.AddRange(GenerateSnapshots((DateTime)report.To));
+
+            this.Context.CommitTransaction();
+
+            return report;
+        }
+
+        /// <summary>
+        /// Update the specified project report metadata in the datasource.
+        /// If the project 'To' date has changed, 
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'report' is required.</exception>
+        /// <exception cref="RowVersionMissingException">Project report rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to edit report.</exception>
+        /// <returns></returns>
+        public ProjectReport Update(ProjectReport report)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var originalReport = this.Context.ProjectReports
+                .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
+
+            if (originalReport.To == null) throw new ArgumentNullException();
```

Comment:
I'm going to resolve this as I think this was covered by other discussion.

Reply:
I don't think we want the entity `To` to be nullable.  Just the models used by the API correct?

Reply:
if only the entity is null, you end up with the same problem as you do when the model is null, which is that you won't get an error from a null/not defined To: value, instead you will get a Date value of: 1/1/0001 12:00:00 AM.

In my mind, the entire point of making a DateTime not nullable is that you should get an error if the API caller or automapper doesn't provide a value.

I actually think the alternative is much worse, because before I was using a nullable datetime I ended up with a whole bunch of 1/1/0001 12:00:00 AM saved to the DB.

Reply:
There is a relationship between the model and the entity, but in the DB the `To` value should never be `null`.  Additionally nullable values in a DB are not performant, nor are they considered good practice.  Especially when the value shouldn't be nullable in any case.

The only time I can see the `To` value being null is when a user makes an AJAX request and they are specifying the `To` because the backend is going to.  But the reality is the backend will always set the `To` to the current moment in time, never null.

Reply:
Does that mean that you would be fine if I left the entity as-is if I added this to the DB configuration?

```
builder.Property(m => m.From).HasColumnType("DATETIME2").IsRequired();
builder.Property(m => m.To).HasColumnType("DATETIME2").IsRequired();
```

Reply:
Interesting solution.  Does it correctly handle the table creation?  I suppose if it does it would satisfy the majority of the issues.

Reply:
![image](https://user-images.githubusercontent.com/41091511/96799899-ee6d9000-13b8-11eb-90dc-a2b1573b4585.png)

Reply:
![image](https://user-images.githubusercontent.com/41091511/96799955-10671280-13b9-11eb-9b0a-1956303cb0a7.png)